presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -438,10 +440,59 @@ public ActualProperties visitJoin(JoinNode node, List<ActualProperties> inputPro
+                    // for the join in terms of its outputs
+                    probeProperties = probeProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+                    buildProperties = buildProperties.translate(column -> filterIfMissing(node.getOutputSymbols(), column));
+
+                    if (!SystemSessionProperties.isDeriveFullJoinProperties(session) ||
+                            !probeProperties.getNodePartitioning().isPresent() ||
+                            !buildProperties.getNodePartitioning().isPresent()) {
+                        return ActualProperties.builder()
+                                .global(probeProperties.isSingleNode() ? singleStreamPartition() : arbitraryPartition())
+                                .build();
+                    }
+
+                    checkState(
+                            probeProperties.getNodePartitioning().get().getArguments().size() == buildProperties.getNodePartitioning().get().getArguments().size(),
+                            "Inconsistent partitioning for left and right side of join: %s vs %s",
+                            probeProperties.getNodePartitioning().get(),
+                            buildProperties.getNodePartitioning().get());
+
+                    List<Partitioning.ArgumentBinding> newBindings = new ArrayList<>();
+                    for (int i = 0; i < probeProperties.getNodePartitioning().get().getArguments().size(); i++) {
+                        Partitioning.ArgumentBinding left = probeProperties.getNodePartitioning().get().getArguments().get(i);
+                        Partitioning.ArgumentBinding right = buildProperties.getNodePartitioning().get().getArguments().get(i);
+
+                        if (left.isVariable() && right.isVariable()) {
+                            boolean found = false;
+                            for (JoinNode.EquiJoinClause equality : node.getCriteria()) {
+                                if (equality.getLeft().equals(left.getColumn()) && equality.getRight().equals(right.getColumn()) ||
+                                        equality.getLeft().equals(right.getColumn()) && equality.getRight().equals(left.getColumn())) {
+                                    found = true;
+                                    newBindings.add(Partitioning.ArgumentBinding.expressionBinding(new CoalesceExpression(ImmutableList.of(equality.getLeft().toSymbolReference(), equality.getRight().toSymbolReference()))));
+                                }
+                            }
+                            checkState(found, "Partitioning inconsistent with join criteria");
+                        }
+                        else if (left.isConstant() && right.isVariable()) {
+                            checkState(buildProperties.getConstants().get(right.getColumn()).equals(left.getConstant()));
```

Comment:
so `right` is constant after all? Waat guarantees that? Why `right.isConstant()` is false

Reply:
Yes, but it’s recorded as a variable in the partitioning function call and there’s a constant mapping kept separately in the properties. I know, the way partitioning functions are being modeled is very confusing... we should fix that at some point.